### DISCUSSION TITLE
V12.1+ and Firebase AppCheck

### DISCUSSION THREAD

**Discussion by artecoop** [vercel/next.js#35689] [Help]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

  Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 21.4.0: Mon Feb 21 20:36:53 PST 2022; root:xnu-8020.101.4~2/RELEASE_ARM64_T8101
    Binaries:
      Node: 16.14.2
      npm: 8.5.5
      Yarn: 1.22.18
      pnpm: 6.11.0
    Relevant packages:
      next: 12.0.10
      react: 17.0.2
      react-dom: 17.0.2
      firebase: 9.6.10

### Describe the Bug

I've found a problem using Firebase AppCheck after updating Next to 12.1+

here's how I setup Firebase in my project
in lib/firebase.ts I have:
```
import { initializeApp } from 'firebase/app';
import { getAuth } from 'firebase/auth';
import { initializeAppCheck, ReCaptchaV3Provider } from 'firebase/app-check';

const firebaseApp = initializeApp({
    apiKey: '...',
    authDomain: '...'
});

initializeAppCheck(firebaseApp, {
    provider: new ReCaptchaV3Provider('...'),
    isTokenAutoRefreshEnabled: true
});

const auth = getAuth(firebaseApp);
export default auth;
```

After updating, either running or building, I get this error:
```
info  - Collecting page data ..node:internal/process/promises:279
            triggerUncaughtException(err, true /* fromPromise */);
            ^

ReferenceError: document is not defined
    at makeDiv (file:///Users/xxx/DEV/yyy/node_modules/@firebase/app-check/dist/esm/index.esm2017.js:1044:26)
    at initializeV3 (file:///Users/xxx/DEV/yyy/node_modules/@firebase/app-check/dist/esm/index.esm2017.js:989:19)
    at ReCaptchaV3Provider.initialize (file:///Users/xxx/DEV/yyy/node_modules/@firebase/app-check/dist/esm/index.esm2017.js:1168:9)
    at _activate (file:///Users/xxx/DEV/yyy/node_modules/@firebase/app-check/dist/esm/index.esm2017.js:1451:23)
    at initializeAppCheck (file:///Users/xxx/DEV/yyy/node_modules/@firebase/app-check/dist/esm/index.esm2017.js:1407:5)
    at /Users/xxx/DEV/yyy/public/.next/server/pages/_app.js:607:71 {
  type: 'ReferenceError'
}
```

reverting nextjs to 12.0.10 solve the problem

### Expected Behavior

To work as expected

### To Reproduce

Even with a fresh install of next and firebase with the simple setup explained in "Describe the Bug", without adding any other packages, returns the problem.
Tested with current production release and v12.1.3-canary.0

**Comment by balazsorban44** [comment]
Looks like your code/Firbease references `document` in a non-browser environment. Make sure it's not the case, as Node.js does not have `document`, hence the error.

If your code/library cannot run in Node.js, you can import the component using it dynamically with `ssr: false` so Next.js does not do a server pass on it: https://nextjs.org/docs/advanced-features/dynamic-import#with-no-ssr

**Comment by artecoop** [comment]
Hey @balazsorban44 thanks for your answer.
This is the first time I use dynamic import so I'm a little confused: if I try to dynamically import the firebase ts file from lib I get this error: ` Type 'Auth' is not assignable to type 'ComponentType<{}> | { default: ComponentType<{}>; }'`

Is dynamic import dedicated to components? Because my lib for firebase exports an interface representing Firebase Auth service.

**Comment by artecoop** [comment]
@balazsorban44 No news about?

**Comment by ChazUK** [comment]
I've also found this issue and not 100% sure what the solution is. I want to have my firebase functions exported as a library, and I'm not sure where App Check needs to be called, straight after initialising the app? Before any firestore etc calls are made? So for the moment I have wrapped it to only load in the browser

```javascript
import { getApp, initializeApp } from "firebase/app";
import { initializeAppCheck, ReCaptchaV3Provider } from "firebase/app-check";

initializeApp({ ... });

if (typeof window !== "undefined") {
  appCheck = initializeAppCheck(getApp(), {
    provider: new ReCaptchaV3Provider("..."),
    isTokenAutoRefreshEnabled: true,
  });
}
```

This might also be useful to get NextJS to verify when building the static pages? https://firebase.google.com/docs/app-check/custom-resource-backend

**Reply by jasan-s** [reply]
Doing this leads to an error the sitekey is missing

**Reply by NomadicDeveloper22** [reply]
> Doing this leads to an error the sitekey is missing

Did you replace the `...` ?

**Comment by jasan-s** [comment]
Has anyone been able to get firebase app-check work with Next js?

if i follow the firebase docs i get the following error in next js app 

>Error: Component app-check has not been registered yet

if i add if condition for initializeAppCheck then i get the following error: 
> Error: Missing required parameters: sitekey

**Comment by thatAwesomeKK** [comment]
You cannot to SSR using AppCheck in NEXTJS.
Been scratching my head for so long

**Reply by Judert** [reply]
My "solution" was to just use the admin sdk from the next api and rate limit it. My security rules basically allow no public calls. Kinda defeats the whole purpose of firebase, just going to use the t3 stack for my next project.

**Reply by Judert** [reply]
Even if this app check thing worked it wouldn't prevent a ddos attack or anything as far as I'm aware.

**Reply by NomadicDeveloper22** [reply]
> Even if this app check thing worked it wouldn't prevent a ddos attack or anything as far as I'm aware.

It's not meant to prevent DDOS attacks. It's purpose to verify your data is only requested by your applications

**Comment by vedantmgoyal9** [comment]
@artecoop did you find any solution for this? If not, can you file a bug at the Firebase SDK repo, maybe they could help.

**Reply by artecoop** [reply]
Actually no, i did not find any solution. To be honest, I didn't search for one tho. I left aside implementing AppCheck on my project, then the release of Nextjs 13 convinced me to try other metaframework, hopefully finding a solution to this (and other) problem.

**Reply by vedantmgoyal9** [reply]
Can you file a bug at the Firebase SDK repo, maybe they could help. I think it is an issue in Firebase SDK itself.

**Reply by vedantmgoyal9** [reply]
cc @artecoop

**Reply by artecoop** [reply]
Well, I'm trying to update my codebase with next13 before filing the bug. I'll keep you updated, as soon as I have time to do it

**Comment by juokavom** [comment]
I struggled with this for a few hours, but then with the help of Chat GPT, got a solution which surprised even me, because it worked!

Enjoy:
```javascript
if (typeof window !== "undefined") {
  import("firebase/app-check").then((firebaseAppCheck) => {
    firebaseAppCheck.initializeAppCheck(app, {
      provider: new firebaseAppCheck.ReCaptchaV3Provider(/* YOUR KEY HERE */),
      isTokenAutoRefreshEnabled: true,
    });
  });
}
```

From what I understood there was a problem with Firebase App check library integration with Next.js - the app check library trying to access document would fail as Next.js offers SSR (and that is written in Node.js which is backend based and does not have document), so we need to make sure that window (document) exists and then import the lib.

**Reply by arpankotecha** [reply]
Thank you for sharing this. Your comment helped me to find the solution to my problem. In my case, I was able to import it outside the if statement. I just needed to initialize AppCheck inside the if statement.

**Reply by pierre-phil** [reply]
I'm getting an [error 400 : `app not registered`](https://stackoverflow.com/questions/78210633/problem-with-firebase-app-check-app-is-not-registered-next-js-integration) when implementing App Check this way. Anyone can help ? [See this thread](https://stackoverflow.com/questions/78210633/problem-with-firebase-app-check-app-is-not-registered-next-js-integration)
